ops/model.py

```diff
@@ -392,6 +392,10 @@ def __init__(self, relation_name, relation_id, is_peer, our_unit, backend, cache
                 self.units.add(unit)
                 if self.app is None:
                     self.app = unit.app
+            if self.app is None:
+                app_name = os.environ.get('JUJU_REMOTE_APP')
+                if app_name is not None:
+                    self.app = cache.get(Application, app_name)
```

Comment:
This is the fix for the other issue we were talking about not finding the app?

Please separate this fix in other branch. 

This is too long and it would seriously simplify the review if only tests changes are included (not real functionality change).

Thanks!

Reply:
I think there's a bad merge somewhere because a lot of the changes seem to already be on master

Reply:
I didn't intend to land this with this PR, I had written it down as a prototype of what we might do, but didn't revert it before landing other code. I'll pull this out, and then make sure that I'm up to date with master.